### PR TITLE
bulkio: Make incremental scheduled backup wait for full backup.

### LINKED ISSUE #52835: bulkio: Wait for full backup before executing incremental schedule
Increment scheduled backup should wait for the full backup to complete
before it starts runnig.

### PR DESCRIPTION
**Pull request by miretskiy**
Fixes #52835

Add ability to record schedule groups: set of related schedules.
Use this functionality to makean incremental schedule wait
until the full one completes before it begins its execution.

Release Notes: None

### CODE REVIEW

**Comment by miretskiy** [comment]
@dt -- rebased on master; ready for review

File: pkg/ccl/backupccl/backup_job.go [hunk 1]
```diff
@@ -571,24 +572,45 @@ func (b *backupResumer) Resume(
...
-	info := b.job.CreatedBy()
+	env := scheduledjobs.ProdJobSchedulerEnv
+
+	if err := exec.DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		// Do not rely on b.job containing created_by_id.  Query it directly.
+		datums, err := exec.InternalExecutor.QueryRowEx(
+			ctx,
+			"lookup-schedule-info",
+			txn,
+			sqlbase.InternalExecutorSessionDataOverride{User: security.RootUser},
```
Comment (dt) [inline pkg/ccl/backupccl/backup_job.go:590]: this should probably be `NodeUser`

File: pkg/ccl/backupccl/create_scheduled_backup.go [hunk 2]
```diff
@@ -320,54 +307,69 @@ func doCreateBackupSchedules(
...
 
 		// If needed, create incremental.
 		if incRecurrence != nil {
 			backupNode.AppendToLatest = true
+			inc, err := makeBackupSchedule(
+				env, p.User(), fullScheduleName+": INCREMENTAL", incRecurrence, details, backupNode)
+			if err != nil {
+				return err
+			}
+			inc.Pause("Waiting for FULL")
```
Comment (dt) [inline pkg/ccl/backupccl/create_scheduled_backup.go:341]: "Waiting for initial backup to complete"

**Review by dt — COMMENTED** [review]
Hm. I donno, this might just be me losing my marbles after staring at too much code, so feel free to just say so, but on first read, I'm about dubious of introducing new concepts like "ScheduleGroup" and "GroupAction".  Again, might be just me, but I find the it is a bit harder to grok with the new concepts. Or, alternatively, I expect them to be actual rows in a table called `schedule_groups` and then to have a `group_id` column in schedules or whatever. I also don't know that `GroupAction` is really _general_ to all schedule types?

Thus, my counterproposal would be: 
1) the _backup specific_ `ScheduledBackupExecutionArgs` gains a "optional int64 unpause_on_success` field 
2) the success hook for backups looks at it and, if it is set: a) unpauses the schedule with that ID b) clears it from itself.

WDYT?

**Review by miretskiy — COMMENTED** [review]
Sure -- that makes sense.  I had something like this going -- notify success is a bit tricky w/ jobs; and initially, I was having issues w/ doing it in the resumer and kinda got on a path of doing it through "meta" proto... Eventually, backup resumer on success started working fine - so making these changes backup specific is fine by me.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52836#-:-MF50YqOCUMAIPkDSUMe:bes34qj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)

___
*[pkg/ccl/backupccl/backup_job.go, line 590 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52836#-MEiYs0iyOK0U3ZEuLm5-r1-590:-MF52gFI4ddytWV80sGm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbe70eeffc206304bdccda013f53b2f9bbbc738b/pkg/ccl/backupccl/backup_job.go#L590)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

this should probably be `NodeUser`
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 341 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52836#-MEiYs0iyOK0U3ZEuLm6-r1-341:-MF54vmEEW7ageh63pRO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbe70eeffc206304bdccda013f53b2f9bbbc738b/pkg/ccl/backupccl/create_scheduled_backup.go#L341)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

"Waiting for initial backup to complete"
</blockquote></details>

Done.

**Review by dt — APPROVED** [review]
Looks like you picked up the go 1.14 proto diff

**Comment by miretskiy** [comment]
Argh... Will fix.

**Pull request merged**